File: hIndex/h_index.cpp
// a O(n) complexity method
int hval(const vector<int> H) {  
    deque<int>q;  
    q.push_back(0);  
    int hval = 0;  
    for (auto i : H) {  
  
        if (hval == 0 && i > 0)hval = 1;  
        else if(q.size()>0 &&  (q.size()>=hval) && (q.front()>hval)&&((i>q.back())||(i>hval)))  
            ++hval;   
  
        while (!q.empty() && q.front() <= hval)  
            q.pop_front();  
        if (q.empty() || i > q.back())  
            q.push_back(i);  
        else q.push_front(i);  
        while (q.size() > hval)  
            q.pop_front();  
    }  
    return hval;  
}  

File: hIndex/h_index.py
# this is a method with O(n*log n) complexity
def hindex(citel):  
    citel.sort()# O(n*log n)
    h=1 if citel[-1]>0 else 0  
    small=citel[-1]  
    for i in citel[-2::-1]:  
        if i==small and i >h:  
            h+=1  
        elif i>h:  
            h+=1  
            small=i  
        else:  
            break  
    return h  
